src/scripting/scriptingService.ts

```diff
@@ -148,31 +409,156 @@ export class ScriptingService {
+                return "Alter";
+            default:
+                return "Unknown";
+        }
+    }
+
     public async script(scriptingParams: IScriptingParams): Promise<string> {
-        const result = await this._client.sendRequest(ScriptingRequest.type, scriptingParams);
-        return result.script;
+        const scriptTelemetryActivity = startActivity(
+            TelemetryViews.ScriptingService,
+            TelemetryActions.Script,
+            undefined,
+            {
+                operation: this.stringifyScriptOperation(scriptingParams.operation),
+            },
+        );
+        return await vscode.window.withProgress(
+            {
+                location: vscode.ProgressLocation.Notification,
+                title: LocalizedConstants.ObjectExplorer.GeneratingScript,
+                cancellable: true,
+            },
+            async (_progress, token) => {
+                const scriptPromise = new Deferred<{
+                    script: string;
+                    errorMessage: string;
+                    errorDetails: string;
+                }>();
+
+                let operationId: string | undefined;
+
+                token.onCancellationRequested(() => {
+                    if (!operationId) {
+                        scriptTelemetryActivity.end(ActivityStatus.Canceled);
+                        return;
+                    }
+                    this._client.sendRequest(ScriptingCancelRequest.type, {
+                        operationId,
+                    });
+                    const pending = this._onGoingScriptingOperations.get(operationId);
+                    if (pending) {
+                        this._onGoingScriptingOperations.delete(operationId);
+                        pending.resolve({
+                            script: undefined,
+                            errorMessage: SCRIPT_OPERATION_CANCELED_ERROR,
+                            errorDetails: undefined,
+                        });
+                    }
+                });
+
+                const result = await this._client.sendRequest(
+                    ScriptingRequest.type,
+                    scriptingParams,
+                );
+
+                operationId = result?.operationId;
+
+                if (!operationId) {
+                    const error = new Error("Missing operation id from scripting response");
+                    scriptTelemetryActivity.endFailed(error, false);
+                    scriptTelemetryActivity.endFailed(error, true);
+                    return undefined;
+                }
+
+                this._onGoingScriptingOperations.set(operationId, scriptPromise);
+
+                const scriptResult = await scriptPromise.promise;
+
+                if (scriptResult.errorMessage) {
+                    scriptTelemetryActivity.endFailed(new Error(scriptResult.errorMessage), false);
+                    if (scriptResult.errorMessage === SCRIPT_OPERATION_CANCELED_ERROR) {
```

Comment:
`SCRIPT_OPERATION_CANCELED_ERROR` doesn't need localization because it's never actually displayed to the user, right?